lib/customisation.nix

```diff
@@ -306,4 +310,72 @@ rec {
       };
     in self;
 
+  /* Define a `mkDerivation`-like function based on another `mkDerivation`-like function.
+
+  `mkDerivation` gives access to its final set of derivation attributes when it is passed a function,
+  or when it is passed an overlay-style function in `overrideAttrs`.
+  Instead of composing new `mkDerivation`-like builders using normal function composition,
+  `extendMkDerivation` makes sure that the returned builder supports such
+  first class recursion like `mkDerivation` does.
+
+     Type:
+       extendMkDerivation ::
+         ((AttrSet -> AttrSet) -> Derivation)
+         -> (AttrSet -> AttrSet -> AttrSet)
+         -> (((AttrSet -> AttrSet) | AttrSet) -> Derivation)
+
+     Example:
+       mkLocalDerivation = lib.extendMkDerivation pkgs.stdenv.mkDerivation (finalAttrs:
+         args@{ preferLocalBuild ? true, allowSubstitute ? false, ... }:
+         args // { inherit preferLocalBuild allowSubstitute; })
+
+       mkLocalDerivation.__functionArgs
+       => { allowSubstitute = true; preferLocalBuild = true; }
+
+       mkLocalDerivation { inherit (pkgs.hello) pname version src; }
+       => «derivation /nix/store/fh9jww7kv5v814lasn78ydck7fb6hsjr-hello-2.12.1.drv»
+
+       mkLocalDerivation (finalAttrs: { inherit (pkgs.hello) pname version src; })
+       => «derivation /nix/store/fh9jww7kv5v814lasn78ydck7fb6hsjr-hello-2.12.1.drv»
+  **/
+  extendMkDerivation =
+    # mkDerivation-like builder to extend
+    mkDerivationBase:
+    # Overlay implementation of the new builder
+    impl:
+    lib.copyFunctionArgs
+      # Make the __functionArgs looks like one from a builder accepting plain attribute set.
+      (impl { })
+      # Adds the fixed-point style support
+      (rattrs: (mkDerivationBase (finalAttrs:
+        impl finalAttrs (if lib.isFunction rattrs then rattrs finalAttrs else rattrs)
+      )));
+
+
+  /* Like `extendMkDerivation`, but additionally accepts a function to be applied to the resulting derivation.
+
+     Type:
+       extendMkDerivation ::
+         ((AttrSet -> AttrSet) -> Derivation)
+         -> (Derivation -> Derivation)
+         -> (AttrSet -> AttrSet -> AttrSet)
+         -> (((AttrSet -> AttrSet) | AttrSet) -> Derivation)
+   */
+  extendMkDerivationModified =
+    let
+      # Apply derivation modifiers while taking care of overrideAttrs
+      applyDrvModifier = modifier: drv: modifier drv // {
+        overrideAttrs = fdrv: applyDrvModifier modifier (drv.overrideAttrs fdrv);
```

Comment:
The interaction with `.override` isn't covered here.

(So also needs tests)

Reply:
> The interaction with `.override` isn't covered here.

IIUC, `.override` is added by `callPackages`, which is typically applied after the builder.

Nevertheless, it surely hasn't taken into account the situation in which overriding attributes other than `overrideAttrs` are introduced *by the builder*, no matter if it is `.override` or `.override<Language>Attrs`. It doesn't cause problem in the Python case as we can apply `extendMkDerivation` before adding `overridePythonAttrs`.

Reply:
Oops, I forgot to cover the `.overrideDerivation` attribute.

> The interaction with `.override` isn't covered here.

There is a chance that the result derivation contains the `.override` attribute if the original builder happens to call `callPackage`. In most cases, it will be shadowed by the new `.override` attribute updated when using the new builder to define a package and call it with `callPackage`. I'll get it covered anyway.